### ISSUE TITLE
Enhanced bean can not use final fields in Java 9+.

### ISSUE THREAD

**Issue by soberich** [ebean-orm/ebean#1533]
## Expected behavior
Switching to Java 9+ does not make Ebean to stop working due to 
`Caused by: java.lang.IllegalAccessError: Update to non-static final field .... attempted from a different method (_ebean_setni_...) than the initializer method <init>`
but behaves the same and l=plays nicely with final fields initialized in non-default constructor
## Actual behavior
JPMS was changed in Java 9 to disallow "final" fields from being modified outside of the <init> method. Even other methods called by <init>.
"if the field is final the instruction must occur in the < clinit > method of the current class. Otherwise, an IllegalAccessError is thrown."
And we do really see this on JDK 11:
`java.lang.IllegalAccessError: Update to non-static final field ... attempted from a different method (_ebean_setni_...) than the initializer method <init>`
### Steps to reproduce

```java
package com.example;

import com.example.ScoringRuleEnum;

import javax.persistence.Embeddable;
import java.io.Serializable;

/**
 * @author <a href="mailto:S.Nasin@prof-schumann.ru">Serge Nazin</a> on 17-Oct-18
 */
@Embeddable
public final class ScoringRule implements Serializable {
    private static final long serialVersionUID = 1L;

    private final ScoringRuleEnum    type;

    public ScoringRule(ScoringRuleEnum type) {
        this.type = type;
    }
   public ScoringRuleEnum getType() {
        return type;
    }
}
```
generates... (some generated methods could be omitted)
```java
public final class ScoringRule implements Serializable, EntityBean {
    private static final long                serialVersionUID = 1L;
    private final        ScoringRuleEnum     type;
    private static       String              _EBEAN_MARKER;
    public static        String[]            _ebean_props     = new String[]{"type"};
    protected            EntityBeanIntercept _ebean_intercept = new EntityBeanIntercept(this);
    protected transient  Object              _ebean_identity;

    public ScoringRule(ScoringRuleEnum type) {
        this._ebean_set_type(type);
    }

    public ScoringRuleEnum getType() {
        return this._ebean_get_type();
    }

    public ScoringRule() {
    }

    public String _ebean_getMarker() {
        return _EBEAN_MARKER;
    }

    public String[] _ebean_getPropertyNames() {
        return _ebean_props;
    }

    public String _ebean_getPropertyName(int pos) {
        return _ebean_props[pos];
    }

    public EntityBeanIntercept _ebean_getIntercept() {
        return this._ebean_intercept;
    }

    public EntityBeanIntercept _ebean_intercept() {
        if (this._ebean_intercept == null) {
            this._ebean_intercept = new EntityBeanIntercept(this);
        }

        return this._ebean_intercept;
    }

    protected ScoringRuleEnum _ebean_get_type() {
        this._ebean_intercept.preGetter(0);
        return this.type;
    }

    protected void _ebean_set_type(ScoringRuleEnum newValue) {
        this._ebean_intercept.preSetter(true, 0, this._ebean_get_type(), newValue);
        this.type = newValue;
    }

    protected ScoringRuleEnum _ebean_getni_type() {
        return this.type;
    }

    protected void _ebean_setni_type(ScoringRuleEnum _newValue) {
        this.type = _newValue;
        this._ebean_intercept.setLoadedProperty(0);
    }

    public Object _ebean_getField(int index) {
        if (index == 0) {
            return this.type;
        }
        throw new RuntimeException("Invalid index " + index);
    }

    public Object _ebean_getFieldIntercept(int index) {
        if (index == 0) {
            return this._ebean_get_type();
        }
        throw new RuntimeException("Invalid index " + index);
    }

    public void _ebean_setField(int index, Object o) {
        if (index == 0) {
            this._ebean_setni_type((ScoringRuleEnum) o);
            return;
        }
        throw new RuntimeException("Invalid index " + index);
    }

    public void _ebean_setFieldIntercept(int index, Object o) {
       // Here it was switch I just changed that. It is not releated
        if (index == 0) {
            this._ebean_set_type((ScoringRuleEnum) o);
            return;
        }
        throw new RuntimeException("Invalid index " + index);
    }

    public void _ebean_setEmbeddedLoaded() {
    }

    public boolean _ebean_isEmbeddedNewOrDirty() {
        return false;
    }

    public Object _ebean_newInstance() {
        return new ScoringRule();
    }
}

```
One of possible variants would be to change constructor and generated methods like this
```java
    public ScoringRule(ScoringRuleEnum type) {
        this.type = this._ebean_set_type(type) ;  // HERE ADDED
//        this._ebean_set_type(type);                  // HERE REMOVED
    }
    // HERE was `void`
    protected ScoringRuleEnum _ebean_set_type(ScoringRuleEnum newValue) {
        this._ebean_intercept.preSetter(true, 0, this._ebean_get_type(), newValue);
        return newValue;           // HERE ADDED
//        this.type = newValue; // HERE REMOVED
    }
```

```console
  // some logging output
Caused by: java.lang.IllegalAccessError: Update to non-static final field com.example.ScoringRule.type attempted from a different method (_ebean_setni_type) than the initializer method <init>
        at com.example.ScoringRule._ebean_setni_type(ScoringRule.java:1)
        at com.example.ScoringRule._ebean_setField(ScoringRule.java:1)
        at io.ebeaninternal.server.properties.EnhanceBeanPropertyAccess$Setter.set(EnhanceBeanPropertyAccess.java:79)
        at io.ebeaninternal.server.deploy.BeanProperty.setValue(BeanProperty.java:757)
        at io.ebeaninternal.server.deploy.id.IdBinderSimple.readSet(IdBinderSimple.java:204)
        at io.ebeaninternal.server.query.SqlTreeNodeBean.load(SqlTreeNodeBean.java:265)
        at io.ebeaninternal.server.query.CQuery.readNextBean(CQuery.java:456)
        at io.ebeaninternal.server.query.CQuery.hasNext(CQuery.java:544)
        at io.ebeaninternal.server.query.CQuery.readCollection(CQuery.java:575)
        at io.ebeaninternal.server.query.CQueryEngine.findMany(CQueryEngine.java:394)
        at io.ebeaninternal.server.query.DefaultOrmQueryEngine.findMany(DefaultOrmQueryEngine.java:135)
        at io.ebeaninternal.server.core.OrmQueryRequest.findList(OrmQueryRequest.java:452)
        at io.ebeaninternal.server.core.DefaultServer.findList(DefaultServer.java:1571)
        at io.ebeaninternal.server.core.DefaultServer.findList(DefaultServer.java:1549)
        at io.ebeaninternal.server.query.LimitOffsetPagedList.getList(LimitOffsetPagedList.java:64)
// omitted
```

**Comment by rbygrave** [comment]
I believe the correct thing to do is for the enhancement to remove the `final` modifier.  I'll ponder it some more but that is what I have been thinking we should do.

Cheers, Rob.

**Comment by soberich** [comment]
That's what I did as solution for now. Rob, I am not sure, do you mean it all entity beans from now on should be without a `final` for enhancement, or do you mean that "for now move it away to make it work, and sometime ebean transformer would be reworked accordingly" ?

**Comment by rbygrave** [comment]
I mean all entity bean [fields] from now on should not be final.

Ebean supports partial objects and we often desire many fields to be final (conceptually final at least - from a developer view only setting these fields via a constructor) - so many fields and not just the `@Id` property.  

Ebean must deal with the case where we have a partial and need to additionally load some fields that are "conceptually final" (and Ebean will need to set these loaded fields in a method that isn't the constructor).  

That is, we really can't just use the constructor to support the case of (lazy) loading a partial where some of the properties loaded are these "conceptually final" fields.

In terms of the Java memory model and final fields and "safe construction" this shouldn't be an issue for entity beans per say as they are generally not concurrently shared across threads (at least mutable entity beans are not shared across threads - we actually do have a special case in L2 caching where we allow shared use of read only entity beans - rare super special case). 

So yes, I mean the bytecode enhancement will effectively remove the final modifier from fields.  The developer would see final (or Kotlin val etc) but under the hood Ebean transformation removes the final modifier (and we deem that ok for entity beans from a java memory model perspective).

**Comment by rbygrave** [comment]
BTW: This is a good issue to address fairly urgently with Java 11 out now etc so yes we need to get onto this.

**Comment by rbygrave** [comment]
Fixed with ebean-agent 11.25.1
- refer to https://github.com/ebean-orm/ebean-agent/issues/86

Version 11.25.1 of ebean-agent , ebean-maven-plugin and ebean-gradle-plugin

Fixed in version 7.1 of the maven tile:
```xml
      <plugin>
        <groupId>io.repaint.maven</groupId>
        <artifactId>tiles-maven-plugin</artifactId>
        <version>2.12</version>
        <extensions>true</extensions>
        <configuration>
          <tiles>
            <tile>io.ebean.tile:enhancement:7.1</tile>
          </tiles>
        </configuration>
      </plugin>

```
